lucet-runtime/lucet-runtime-internals/src/region/uffd.rs

```diff
@@ -0,0 +1,512 @@
+                let fault_addr = fault_addr as usize;
+                let fault_page = fault_addr - (fault_addr % host_page_size());
+                let instance_size = limits.total_memory_size();
+
+                let in_region = fault_addr >= start as usize
+                    && fault_addr < start as usize + instance_size * instance_capacity;
+                lucet_ensure!(in_region, "fault is within the uffd region");
+
+                let fault_offs = fault_addr - start as usize;
+                let fault_base = fault_offs - (fault_offs % instance_size);
+                let inst_base = start as usize + fault_base;
+
+                // NB: we are blatantly lying to the compiler here! the lifetime is *not* actually
+                // static, but for the purposes of reaching in to read the sparse page data and the
+                // heap layout, we can treat it as such. The important property to maintain is that
+                // the *real* region lifetime (`'r`) lives at least as long as this handler thread,
+                // which can be shown by examining the `drop` method of `UffdRegion`.
+
+                let inst: &mut Instance = unsafe {
+                    (inst_base as *mut Instance)
+                        .as_mut()
+                        .ok_or(lucet_format_err!("instance pointer is non-null"))?
+                };
+                if !inst.valid_magic() {
+                    eprintln!(
+                        "instance magic incorrect, fault address {:p}",
+                        fault_addr as *mut c_void
+                    );
+                    lucet_bail!("instance magic incorrect");
+                }
+
+                let loc = inst.alloc().addr_location(fault_addr as *const c_void);
+                match loc {
+                    AddrLocation::InaccessibleHeap | AddrLocation::StackGuard => {
+                        // eprintln!("fault in heap guard!");
+                        // page fault occurred out of bounds; trigger a fault by waking the faulting
+                        // thread without copying or zeroing
+                        uffd.wake(fault_page as *mut c_void, host_page_size())
+                            .map_err(|e| Error::InternalError(e.into()))?;
+                    }
+                    AddrLocation::SigStackGuard | AddrLocation::Unknown => {
+                        tracing::error!("UFFD pagefault at fatal location: {:?}", loc);
+                        uffd.wake(fault_page as *mut c_void, host_page_size())
+                            .map_err(|e| Error::InternalError(e.into()))?;
+                    }
+                    AddrLocation::Globals | AddrLocation::SigStack => {
+                        tracing::error!("UFFD pagefault at unexpected location: {:?}", loc);
+                        uffd.wake(fault_page as *mut c_void, host_page_size())
+                            .map_err(|e| Error::InternalError(e.into()))?;
```

Comment:
One of the things I've been seeing in the mmap-related PRs is a re-examination of errors like these.  If there had been a UFFD pagefault at an unexpected location, should the lucet runtime return its generic "InternalError" or should it just panic?  

See related PR: https://github.com/bytecodealliance/lucet/pull/486

Reply:
This is a good call, it should just panic. I think @acfoltzer wanted to take care of that in a second PR. Though tbh I think we should just do it here.